### ISSUE TITLE
Stresstest failures on v2018.06.179.g.9.dce.63318

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#2071]
Don't know which commit they started to fail on, but I can consistently reproduce the failures below on v2018.06.179.g.9.dce.63318.

In addition to those failures, looks like `t/spec/S17-promise/lock-async-stress2.t` segfaults occasionally. Segfaulted twice for me (harness said "Non-zero wait status: 11").

This is on 24-core Debian VM.

```
Test Summary Report
-------------------
t/spec/S04-phasers/in-loop.rakudo.moar                          (Wstat: 256 Tests: 20 Failed: 1)
  Failed test:  14
  Non-zero exit status: 1
t/spec/S17-promise/lock-async-stress.t                          (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 1 tests but ran 0.
Files=1297, Tests=153425, 161 wallclock secs (21.17 usr  3.37 sys + 3492.60 cusr 161.87 csys = 3679.01 CPU)
Result: FAIL
Makefile:585: recipe for target 'm-stresstest5' failed
make: *** [m-stresstest5] Error 1
The spawned command 'make' exited unsuccessfully (exit code: 2)
  in block <unit> at /home/cpan/zscript/bin/z line 154
cpan@perlbuild4~/R/rakudo (master)$ z t t/spec/S04-phasers/in-loop.t t/spec/S17-promise/lock-async-stress.t
t/spec/S04-phasers/in-loop.rakudo.moar .. WARNINGS for /home/cpan/R/rakudo/t/spec/S04-phasers/in-loop.rakudo.moar:
Useless use of constant integer 42 in sink context (line 166)
1..20
not ok 1 - trait blocks work properly in for loop # TODO NEXT/LEAVE ordering RT #124952
# Failed test 'trait blocks work properly in for loop'
# at t/spec/S04-phasers/in-loop.rakudo.moar line 37
# expected: '(1 F)(1 E)(1 a)(1 N)(1 Lv)(2 E)(2 a)(2 b)(2 N)(2 Lv)(3 E)(3 a)(3 N)(3 Lv)(4 E)(4 Lv)(4 Lst)'
#      got: '(1 F)(1 E)(1 a)(1 N)(1 Lv)(2 E)(2 a)(2 b)(2 Lv)(2 N)(3 E)(3 a)(3 N)(3 Lv)(4 E)(4 Lv)(4 Lst)'
not ok 2 - trait blocks work properly in for loop # TODO NEXT/LEAVE ordering RT #124952
# Failed test 'trait blocks work properly in for loop'
# at t/spec/S04-phasers/in-loop.rakudo.moar line 68
# expected: '(1 F1)(1 F2)(1 E1)(1 E2)(1 a)(1 N2)(1 N1)(1 Lv2)(1 Lv1)(2 E1)(2 E2)(2 a)(2 b)(2 N2)(2 N1)(2 Lv2)(2 Lv1)(3 E1)(3 E2)(3 a)(3 N2)(3 N1)(3 Lv2)(3 Lv1)(4 E1)(4 E2)(4 Lv2)(4 Lv1)(4 Lst2)(4 Lst1)'
#      got: '(1 F1)(1 F2)(1 E1)(1 E2)(1 a)(1 N2)(1 N1)(1 Lv2)(1 Lv1)(2 E1)(2 E2)(2 a)(2 b)(2 Lv2)(2 Lv1)(2 N2)(2 N1)(3 E1)(3 E2)(3 a)(3 N2)(3 N1)(3 Lv2)(3 Lv1)(4 E1)(4 E2)(4 Lv2)(4 Lv1)(4 Lst1)(4 Lst2)'
ok 3 - LAST does not fire for empty loop
ok 4 - "last" statement called by LEAVE breaks out of while loop
ok 5 - LEAVE in while loop works as expected
ok 6 - no UnwindException with "last" and "ENTER" in for loop
ok 7 - LAST phaser without block does not put Mu in the iteration variable
ok 8 - LAST phaser with    block does not put Mu in the iteration variable
ok 9 - FIRST in loop works fine with recursion
ok 10 - NEXT in loop works fine with recursion
ok 11 - LAST in loop works fine with recursion
ok 12 - NEXT can see outer $_
ok 13 - LAST can see outer $_
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 169
not ok 14 - POST should not see outer $_
# Failed test 'POST should not see outer $_'
# at t/spec/S04-phasers/in-loop.rakudo.moar line 167
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in code  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 177
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
Use of uninitialized value of type Mu in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at /home/cpan/R/rakudo/lib/Test.pm6 (Test) line 184
# expected: '42 42 42 42 42 42 42 42 42 42'
#      got: '         '
Use of uninitialized value of type Any in string context.
Methods .^name, .perl, .gist, or .say can be used to stringify it to something meaningful.
  in block  at t/spec/S04-phasers/in-loop.rakudo.moar line 174
not ok 15 - KEEP should not see outer $_ # TODO KEEP should not see outer 
# Failed test 'KEEP should not see outer $_'
# at t/spec/S04-phasers/in-loop.rakudo.moar line 174
# expected: '42 42 42 42 42 42 42 42 42 42'
#      got: ''
ok 16 - LAST can see outer $_
ok 17 - LEAVE can see outer $_
ok 18 - PRE can see outer $_
ok 19 - UNDO can see outer $_
    1..2
    ok 1 - unwanted loop
    ok 2 - wanted loop
ok 20 - FIRST+LAST loops as last statement in subs work and do not crash
# FUDGED!
# Looks like you failed 1 test of 20
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/20 subtests 
t/spec/S17-promise/lock-async-stress.t .. 
1..1
ok 1 - Lock provides mutual exclusion when pushing to array
ok

Test Summary Report
-------------------
t/spec/S04-phasers/in-loop.rakudo.moar (Wstat: 256 Tests: 20 Failed: 1)
  Failed test:  14
  Non-zero exit status: 1
Files=2, Tests=21, 44 wallclock secs ( 0.03 usr  0.00 sys + 89.54 cusr  0.47 csys = 90.04 CPU)
Result: FAIL
The spawned command 't/fudgeandrun' exited unsuccessfully (exit code: 1)
  in block <unit> at /home/cpan/zscript/bin/z line 154

cpan@perlbuild4~/R/rakudo (master)$ 
```

**Comment by AlexDaniel** [comment]
@dogbert17++ bisected one of the issues:
```perl6
my @a = []; for ^10 { POST @a.push($_); 42 }; say @a
```
```
¦«359ac12d9^»:
WARNINGS for /tmp/tAxhEucWNY:
[42 42 42 42 42 42 42 42 42 42]
Useless use of constant integer 42 in sink context (line 1)

¦«359ac12»:
WARNINGS for /tmp/tAxhEucWNY:
[(Mu) (Mu) (Mu) (Mu) (Mu) (Mu) (Mu) (Mu) (Mu) (Mu)]
Useless use of constant integer 42 in sink context (line 1)
```

Bisected: (2018-07-13) https://github.com/rakudo/rakudo/commit/359ac12d9753fc87d2a50dcb06bf387acd4f3de3

Bumped NQP changes: https://github.com/perl6/nqp/compare/2018.06-43-g668be5c...2018.06-50-gb1e80f0

Bumped Moar changes: https://github.com/MoarVM/MoarVM/compare/2018.06-141-gcae4731...2018.06-203-g1b1edfe

**Comment by zoffixznet** [comment]
I'm getting more failures now, with the new files dying with either `Non-zero wait status: 11` or `Non-zero wait status: 7` or not dying at all.

```
Test Summary Report
-------------------
t/spec/S04-phasers/in-loop.rakudo.moar                          (Wstat: 256 Tests: 20 Failed: 1)
  Failed test:  14
  Non-zero exit status: 1
t/spec/S17-promise/lock-async-stress.t                          (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 1 tests but ran 0.
t/spec/S17-promise/lock-async-stress2.t                         (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 1 tests but ran 0.
t/spec/S09-typed-arrays/native-int.rakudo.moar                  (Wstat: 11 Tests: 1631 Failed: 0)
  Non-zero wait status: 11
  Parse errors: Bad plan.  You planned 1633 tests but ran 1631.
t/spec/S06-currying/positional.t                                (Wstat: 11 Tests: 185 Failed: 0)
  Non-zero wait status: 11
  Parse errors: Bad plan.  You planned 232 tests but ran 185.
t/spec/S32-num/stress.t                                         (Wstat: 11 Tests: 1 Failed: 0)
  Non-zero wait status: 11
  Parse errors: Bad plan.  You planned 2 tests but ran 1.
t/spec/integration/eval-and-threads.t                           (Wstat: 11 Tests: 0 Failed: 0)
  Non-zero wait status: 11
  Parse errors: Bad plan.  You planned 1 tests but ran 0.
Files=1299, Tests=153365, 165 wallclock secs (21.28 usr  3.31 sys + 3578.31 cusr 166.99 csys = 3769.89 CPU)
Result: FAIL
Makefile:585: recipe for target 'm-stresstest5' failed
make: *** [m-stresstest5] Error 1
The spawned command 'make' exited unsuccessfully (exit code: 2)
  in block <unit> at /home/cpan/zscript/bin/z line 154
```

**Comment by zoffixznet** [comment]
Also, I saw at least one docs repo travis job die with a SEGV: https://travis-ci.org/perl6/doc/jobs/404390772

**Comment by zoffixznet** [comment]
More info in https://github.com/MoarVM/MoarVM/issues/907 and https://github.com/MoarVM/MoarVM/issues/906

**Comment by zoffixznet** [comment]
Updated list of stresstest failures after [latest nqp/MoarVM bump](https://github.com/rakudo/rakudo/commit/f9743ad82482d4d7dad1a1b1fd4ce1877acbe7ee):

```
$ ./perl6 -v
This is Rakudo version 2018.06-226-gf9743ad built on MoarVM version 2018.06-265-gfdd5719



Test Summary Report
-------------------
t/spec/S04-phasers/in-loop.rakudo.moar                          (Wstat: 256 Tests: 20 Failed: 1)
  Failed test:  14
  Non-zero exit status: 1
t/spec/S17-promise/lock-async-stress2.t                         (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 1 tests but ran 0.
t/spec/S17-promise/lock-async-stress.t                          (Wstat: 11 Tests: 0 Failed: 0)
  Non-zero wait status: 11
  Parse errors: Bad plan.  You planned 1 tests but ran 0.
t/spec/S17-supply/syntax.t                                      (Wstat: 256 Tests: 74 Failed: 0)
  Non-zero exit status: 1
  Parse errors: Bad plan.  You planned 83 tests but ran 74.
Files=1300, Tests=153416, 176 wallclock secs (26.01 usr  3.74 sys + 3850.50 cusr 170.65 csys = 4050.90 CPU)
```

**Comment by jnthn** [comment]
`t/spec/S04-phasers/in-loop.rakudo.moar` was caused by code-gen improvements so that we called the body of a loop in void context. That meant, however, that we lost access to the loop body result and could not call the exit handler with it. Resolved in https://github.com/MoarVM/MoarVM/commit/8d96c86c78.

**Comment by lizmat** [comment]
FWIW, I don't see that one anymore

**Comment by zoffixznet** [comment]
```
All tests successful.
Files=1301, Tests=153427, 167 wallclock secs (22.09 usr  3.36 sys + 3623.73 cusr 182.31 csys = 3831.49 CPU)
Result: PASS
```

**Comment by jnthn** [comment]
The only one I see happening occasionally now on Rakudo HEAD is `S17-promise/lock-async-stress2.t`. In https://github.com/rakudo/rakudo/commit/bf514c1502 I fixed a bug that leads to quite a lot less memory pressure in those tests, and I can't get S17-promise/lock-async-stress.t to fail at all any more (plus it runs quite a bit faster!)

It seems that there's some kind of heisenbug afflicting the async lock tests, but it goes back a long time. It also got a lot less common once I ran without optimizations, suggesting that 1) it's a race, and 2) we're maybe noticing it more now because other things got faster and made the race more likely.

I'll continue hunting it in the future, of course, but it's not a regression, so I don't think it has implications for this release.
